zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -386,6 +435,99 @@ where
             }
         }
 
+        #[cfg(feature = "transparent-inputs")]
+        struct Zip320SecondStep {
+            transaction_request: TransactionRequest,
+            payment_pools: BTreeMap<usize, PoolType>,
+            prior_step_inputs: StepOutput,
+            balance: TransactionBalance,
+        }
+
+        // If ephemeral state was detected, handle it here.
+        #[cfg(feature = "transparent-inputs")]
+        let (transaction_request, second_step) = if ephemeral_outputs.is_empty() {
+            (transaction_request, None)
+        } else {
+            // Construct a new TransactionRequest from `transaction_request` that excludes
+            // the ephemeral outputs, and in their place includes a single transparent
+            // output to the legacy address for this account.
+
+            let mut proposal1_excludes = BTreeSet::new();
+            let mut proposal2_payments = vec![];
+            let mut proposal2_payment_pools = BTreeMap::new();
+            let mut total_ephemeral_plus_fee: NonNegativeAmount = (|| todo!())();
+            let (first_idx, _) = ephemeral_outputs[0];
+            for (idx2, (idx1, payment)) in ephemeral_outputs.into_iter().enumerate() {
+                total_ephemeral_plus_fee = (total_ephemeral_plus_fee + payment.amount).ok_or(
+                    InputSelectorError::Selection(GreedyInputSelectorError::Balance(
+                        BalanceError::Overflow,
+                    )),
+                )?;
+                proposal1_excludes.insert(idx1);
+                proposal2_payments.push(payment);
+                proposal2_payment_pools.insert(idx2, PoolType::Transparent);
+            }
+
+            // TODO: Calculate fee for transaction request 2 and add it to
+            // total_ephemeral_plus_fee.
+
+            let proposal1_transaction_request = TransactionRequest::from_indexed(
+                transaction_request
+                    .payments()
+                    .iter()
+                    .filter_map(|(idx, payment)| {
+                        // If this is the first ephemeral output in the original request,
+                        // replace it with a transparent output to the legacy address.
+                        // We use the legacy address here only as a dummy address for the
+                        // ZIP-321 URI serialized in the proposal. It will be replaced by
+                        // a unique ephemeral P2PKH address when the proposal is created.
```

Comment:
An alternate approach:

We should really be generating a gap-limit sized batch of unused transparent addresses for detection of payments that might have gone to another wallet anyway, and marking them as used once we've detected a payment to that address. So, what if we just did that, and then picked the first of those addresses here?

Reply:
The issue is that there is no restriction on how many times the wallet creates a proposal without necessarily sending it. If we choose the ephemeral address at the time of making the proposal, then a fresh addresses must be used for each one, and so there is no way to enforce the gap limit. If the proposal only references some kind of token or a dummy address, on the other hand, then it can be replaced by a unique real address when the transactions are created. Once we create a transaction we will successfully send it unless there is some error, and so it's much less likely that the gap limit will be violated.

Reply:
> each time the wallet constructs a proposal, then it will take the first unused address from the gap limit queue. It will just pick the same address every time until the transaction is actually constructed

That is what I originally suggested, but in discussion with @daira and @ebfull 

> we don't do anything to lock notes or UTXOs as part of proposal construction.

Indeed, but that doesn't mean we avoid ephemeral address reuse (for example, once #1260 is fixed, the caller could create two proposals, one funded from Sapling and the other funded from Orchard, and these would not have colliding input notes and could co-exist).

If a proposal is generated with the ephemeral address included, then that ephemeral address has been revealed; this is doubly true given that (per discussions I recall having with you) proposals are designed to be serializable beyond an internal FFI, and so even if a proposal is not used we can't assume it doesn't escape containment.